programs/mango-v4/src/state/orderbook/book.rs

```diff
@@ -128,6 +129,52 @@ impl<'a> Orderbook<'a> {
             remaining_quote_lots -= match_quote_lots;
             assert!(remaining_quote_lots >= 0);
 
+            let self_trade_behavior = match order.params {
+                OrderParams::Market {
+                    self_trade_behavior,
+                } => Some(self_trade_behavior),
+                OrderParams::ImmediateOrCancel {
+                    self_trade_behavior,
+                    ..
+                } => Some(self_trade_behavior),
+                _ => None,
+            };
+
+            let mut maker_fee = market.maker_fee;
+            let mut taker_fee = market.taker_fee;
+            let order_would_self_trade = mango_account_pk.eq(&best_opposing.node.owner);
+            if order_would_self_trade {
+                if let Some(self_trade_behavior) = self_trade_behavior {
+                    require!(
+                        self_trade_behavior != SelfTradeBehavior::AbortTransaction,
+                        MangoError::WouldSelfTrade
+                    );
+
+                    if self_trade_behavior == SelfTradeBehavior::CancelProvide {
```

Comment:
would prefer to `match {}` on self_trade_behavior to guarantee the check is exhaustive